### PR TITLE
Allow full and relative path wallet database location

### PR DESCRIPTION
**Pull request by oxarbitrage**
Closes https://github.com/zcash/zcash/issues/1873

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -858,9 +858,20 @@ bool CWallet::Verify()
     LogPrintf("Using wallet %s\n", walletFile);
     uiInterface.InitMessage(_("Verifying wallet..."));
 
-    // Wallet file must be a plain filename without a directory
-    if (walletFile != boost::filesystem::basename(walletFile) + boost::filesystem::extension(walletFile))
-        return UIError(strprintf(_("Wallet %s resides outside data directory %s"), walletFile, GetDataDir().string()));
+    if (walletFile != boost::filesystem::basename(walletFile) + boost::filesystem::extension(walletFile)) {
+        boost::filesystem::path path(walletFile);
+        if(path.is_absolute()) {
```
Comment (str4d) [inline src/wallet/wallet.cpp:863]:
```suggestion
        if (path.is_absolute()) {
```

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -858,9 +858,20 @@ bool CWallet::Verify()
...
+            if (!boost::filesystem::exists(path.parent_path())) {
+                return UIError(strprintf(_("Absolute path %s do not exist"), walletFile));
+            }
+        }
+        else {
```
Comment (str4d) [inline src/wallet/wallet.cpp:868]:
```suggestion
        } else {
```

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -858,9 +858,20 @@ bool CWallet::Verify()
...
+            boost::filesystem::path full_path = GetDataDir() / path;
+            if(!boost::filesystem::exists(full_path.parent_path())) {
```
Comment (str4d) [inline src/wallet/wallet.cpp:870]:
```suggestion
            if (!boost::filesystem::exists(full_path.parent_path())) {
```

**Review by str4d — APPROVED** [review]
Needs a rebase to fix the merge conflict, and there are some minor whitespace issues; otherwise utACK.

This change does (intentionally) enable writing the `wallet.dat` to anywhere that the user running `zcashd` can write to. I don't think there's a security concern here, as `-wallet` is a configuration or on-start argument, not a runtime decision (unlike e.g. `z_exportwallet`, which we require to specify a filename and not a full path, as the RPC could be accessed remotely).

File: qa/rpc-tests/feature_walletfile.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        assert os.path.isfile(os.path.join(self.options.tmpdir, "node0", "regtest", "altwallet.dat"))
+
+        # test wallet file outside datadir
+        tempname = os.path.join(self.options.tmpdir, "outsidewallet.dat")
+        stop_node(self.nodes[0], 0)
+        self.nodes[0] = start_node(0, self.options.tmpdir, ["-wallet=%s" % tempname])
+        assert os.path.isfile(tempname)
+
+        # full path do not exist
+        invalidpath = os.path.join("/foo/", "foo.dat")
```
Comment (daira) [inline qa/rpc-tests/feature_walletfile.py:34]:
```suggestion
        invalidpath = os.path.join("/this_directory_must_not_exist/", "foo.dat")
```

File: qa/rpc-tests/feature_walletfile.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        self.nodes[0] = start_node(0, self.options.tmpdir, ["-wallet=altwallet.dat"])
+        assert os.path.isfile(os.path.join(self.options.tmpdir, "node0", "regtest", "altwallet.dat"))
+
+        # test wallet file outside datadir
+        tempname = os.path.join(self.options.tmpdir, "outsidewallet.dat")
+        stop_node(self.nodes[0], 0)
+        self.nodes[0] = start_node(0, self.options.tmpdir, ["-wallet=%s" % tempname])
+        assert os.path.isfile(tempname)
+
+        # full path do not exist
```
Comment (daira) [inline qa/rpc-tests/feature_walletfile.py:33]:
```suggestion
        # test the case where absolute path does not exist
        assert !os.path.isdir("/this_directory_must_not_exist")
```

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -873,9 +873,19 @@ bool CWallet::Verify()
...
     uiInterface.InitMessage(_("Verifying wallet..."));
 
-    // Wallet file must be a plain filename without a directory
-    if (walletFile != boost::filesystem::basename(walletFile) + boost::filesystem::extension(walletFile))
-        return UIError(strprintf(_("Wallet %s resides outside data directory %s"), walletFile, GetDataDir().string()));
+    if (walletFile != boost::filesystem::basename(walletFile) + boost::filesystem::extension(walletFile)) {
+        boost::filesystem::path path(walletFile);
+        if (path.is_absolute()) {
+            if (!boost::filesystem::exists(path.parent_path())) {
+                return UIError(strprintf(_("Absolute path %s do not exist"), walletFile));
```
Comment (daira) [inline src/wallet/wallet.cpp:880]:
```suggestion
                return UIError(strprintf(_("Absolute path %s does not exist"), walletFile));
```

File: qa/rpc-tests/feature_walletfile.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        invalidpath = os.path.join("/foo/", "foo.dat")
+        stop_node(self.nodes[0], 0)
+        assert_start_raises_init_error(0, "-wallet=%s" % invalidpath, 
+            "Error: Absolute path %s do not exist")
```
Comment (daira) [inline qa/rpc-tests/feature_walletfile.py:37]:
```suggestion
            "Error: Absolute path %s does not exist")
```

File: qa/rpc-tests/feature_walletfile.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+        # relative path do not exist
+        invalidpath = os.path.join("wallet", "foo.dat")
+        assert_start_raises_init_error(0, "-wallet=%s" % invalidpath, 
+            "Error: Relative path %s do not exist")
```
Comment (daira) [inline qa/rpc-tests/feature_walletfile.py:42]:
```suggestion
            "Error: Relative path %s does not exist")
```

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -873,9 +873,19 @@ bool CWallet::Verify()
...
+            }
+        } else {
+            boost::filesystem::path full_path = GetDataDir() / path;
+            if (!boost::filesystem::exists(full_path.parent_path())) {
+                return UIError(strprintf(_("Relative path %s do not exist"), walletFile));
```
Comment (daira) [inline src/wallet/wallet.cpp:885]:
```suggestion
                return UIError(strprintf(_("Relative path %s does not exist"), walletFile));
```

**Review by daira — APPROVED** [review]
utACK modulo comments.

**Review by str4d — APPROVED** [review]
re-ACK

**Comment by str4d** [comment]
@zkbot r+

**Comment by daira** [comment]
Rebased and force-pushed to correct a bug in my change to the test.

**Comment by daira** [comment]
@zkbot r+

**Pull request merged**